Iss 516 object on surface sampler for complex scenes (#577)
DLR-RM/BlenderProc: examples/datasets/front_3d_object_sampling/README.md

# Sampling objects in the 3D Front Dataset

<p align="center">
<img src="../../../images/front_3d_object_sampling.jpg" alt="Front readme image" width=375>
</p>

> In this example we explain to you how to sample specific objects in the 3D-Front Dataset with the BlenderProc pipeline.
> Whats special is, that we select the surface-object beforehand and that the sampled object is placed directly on the surface of a table for example. 
> This is an advanced example, make sure that you have executed the basic examples before proceeding to this one.

Reviewer: ```suggestion
In this example, we explain to you how to sample specific objects in the 3D-Front Dataset with the BlenderProc pipeline.
What's special is, that we select the surface object beforehand and that the sampled object is placed directly on the surface of a table for example. 
This is an advanced example, make sure that you have executed the basic examples before proceeding to this one.
```

```

## Steps

* Loads the `.json` file: `bproc.loader.load_front3d`. It creates the rooms and also adds emission shaders to the ceiling and lamps.
* Select the surfaces, where the objects should be sampled on: `bproc.object.slice_faces_with_normals()`.
* Load the object, which should be sampled on the surface: `bproc.loader.load_blend()`.
* Sample the object on the surface: `bproc.object.sample_poses_on_surface()`.
* Run the physics simulation: `bproc.object.simulate_physics_and_fix_final_poses`.
* Check if the objects are still on the surface.
* Adds camera pose to the scene: `bproc.camera.add_camera_pose()`. 
* Renders all set camera poses: `bproc.renderer.render()`.
* Writes the output to .hdf5 containers: `bproc.writer.write_hdf5()`.

## Python file (main.py)

#### Front3DObjectSelection 

```python
# Select the objects, where other objects should be sampled on
for obj in room_objs:
    if "table" in obj.get_name().lower() or "desk" in obj.get_name().lower():
        left_objects.append(obj)
for obj in left_objects:
    # The loop starts with and UndoAfterExecution in order to restore the camera positions for each object

>     with bproc.python.utility.UndoAfterExecution():

Reviewer: this was not updated

        # Select the surfaces, where the object should be sampled on
        droppable_surface = bproc.object.slice_faces_with_normals([obj])
```

> * Specify with objects to use to sample on. In this case its "table" and "desk".
> * Then for each of those objects, get the biggest surface (e. g. top of the table) to sample on.

Reviewer: ```suggestion
* Specify with objects to use to sample on. In this case, it's "table" and "desk".
* Then for each of those objects, the biggest surface is selected (e. g. top of the table) for sampling.
```

#### Front3DObjectSampling 

```python
surface_height_z = np.mean(surface_obj.get_bound_box(), axis=0)[2]
        def sample_pose(obj: bproc.types.MeshObject):
            # Sample the spheres location above the surface
            obj.set_location(bproc.sampler.upper_region(
                objects_to_sample_on=[surface_obj],
                min_height=1,
                max_height=4,
                use_ray_trace_check=False
            ))
            #Randomized rotation of the sampled object
            obj.set_rotation_euler(np.random.uniform([0, 0, 0], [np.pi * 2, np.pi * 2, np.pi * 2]))

dropped_object_list = bproc.object.sample_poses_on_surface(sampling_obj, surface_obj, sample_pose,
                                                                   min_distance=0.1, max_distance=10,
                                                                   check_all_bb_corners_over_surface=False)
```

> * Get the hight of the sampling surface
> * For every surface, the object is placed by `bproc.object.smaple_poses_on_surface`.
> * This also needs a pre-defined `sample_pose()` function, where the sampling surface, as well as the min and maximum sampling hight are given.
> * There also the rotation of the sampled object is done.

Reviewer: ```suggestion
* Get the height of the sampling surface
* For every surface, the object is placed by `bproc.object.sample_poses_on_surface`.
* This also needs a pre-defined `sample_pose()` function, where the sampling surface, as well as the minimum and maximum sampling height, are given.
* It also sets the rotation of the sampled object.
```

#### Front3DObjectSamplingChecks

```python
min_coord_z = np.min([dropped_object.get_local2world_mat() @ conv_to_homogen(vert) for vert in

>                               dropped_object.blender_obj.data.vertices], axis=0)[2]

Reviewer 2: This needs an update
